client/web/src/enterprise/batches/detail/changesets/ChangesetSelectRow.tsx

```diff
@@ -1,47 +1,240 @@
+import classNames from 'classnames'
 import InfoCircleOutlineIcon from 'mdi-react/InfoCircleOutlineIcon'
-import React from 'react'
+import React, { Fragment, useCallback, useMemo, useState } from 'react'
 
-import { isErrorLike } from '@sourcegraph/shared/src/util/errors'
 import { pluralize } from '@sourcegraph/shared/src/util/strings'
 
-import { ErrorAlert } from '../../../../components/alerts'
+import { AllChangesetIDsVariables, Scalars } from '../../../../graphql-operations'
+import { eventLogger } from '../../../../tracking/eventLogger'
+import { queryAllChangesetIDs } from '../backend'
+
+import styles from './ChangesetSelectRow.module.scss'
+import { CreateCommentModal } from './CreateCommentModal'
+import { DetachChangesetsModal } from './DetachChangesetsModal'
+
+/**
+ * Describes a possible action on the changeset list.
+ */
+interface ChangesetListAction {
+    /* The type of action. Used internally. */
+    type: string
+    /* The button label for the action. */
+    buttonLabel: string
+    /* The title in the dropdown menu item. */
+    dropdownTitle: string
+    /* The description in the dropdown menu item. */
+    dropdownDescription: string
+    /* Conditionally display the action based on the given query arguments. */
+    isAvailable: (queryArguments: Omit<AllChangesetIDsVariables, 'after'>) => boolean
+    /**
+     * Invoked when the action is triggered. Either onDone or onCancel need to be called
+     * eventually. Can return a JSX.Element to be rendered adacent to the button (i.e. a modal).
+     */
+    onTrigger: (
+        batchChangeID: Scalars['ID'],
+        changesetIDs: () => Promise<Scalars['ID'][]>,
+        onDone: () => void,
+        onCancel: () => void
+    ) => void | JSX.Element
+}
+
+const availableActions: ChangesetListAction[] = [
```

Comment:
Nit, but is this a case (hahahaha #puns) for CONSTANT_CASE?
 
```suggestion
const AVAILABLE_ACTIONS: ChangesetListAction[] = [
```